lkl: Use POSIX semaphores when available

pscollins opened the pull request:
This should behave identically to the current solution made from
pthread condition variables, perhaps with a small performance boost. The
major benefit of this approach is that tools for analyzing threadsafety
(e.g. Valgrind's Helgrind) will understand that a thread woken up by
sem_up has acquired a lock.

Signed-off-by: Patrick Collins pscollins@google.com

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/lkl/linux/69)

<!-- Reviewable:end -->

pscollins commented:
Also, it might be nice to get more flavors of POSIX hosts on CircleCI in the future (e.g. FreeBSD, Darwin) to make sure that builds stay compatible with all advertised platforms --- I only have my Linux workstation to test on so I worry that I'll accidentally add something that's Linux-specific without realizing it.

pscollins commented:
Ah, but this doesn't work. I guess because of something to do with how `pthread_cond_t`s work regarding `EINTR`? @tavip if you have a moment to explain why `pthread_cond_t`s are needed rather than `sem_t`s I would appreciate it --- I assumed it was for platform compatibility reasons but obviously there's some meaning there that `sem_t`s don't provide, because this hangs and/or hits failed assertions when `sem_wait` returns -1 and sets `EINTR`.

stfairy commented:
Maybe we should put this as a CONFIG_LKL_something?

pscollins commented:
@stfairy I'm not sure why the tests aren't registering as failing but I think this is very broken --- it looks like the major difference is that `pthread_cond_t` never errors out with `EINTR` or `EAGAIN`. With this patch, `sem_wait` usually hits the `assert(!ret)` very quickly and crashes. I tried to test it on my local machine before submitting this patch and it seemed fine, but I guess there was something funny with my build environment, because the last few builds I did with this patch didn't work. The test output on CircleCI is empty so I think that may be happening there, too. 

I tried changing it so that `sem_wait` retries if it errors out with `EINTR` or `EAGAIN`, but then the program hanged and eventually failed with an error about running out of space. If I just let it continue when it hits either of those errors then it eventually fails because it tries to `sem_post` a semaphore it never successfully `sem_wait`ed.

So it looks like the the semantics of `pthread_cond_t` vs `sem_t` are significant here? I'm not clear how `pthread_cond_t` reacts in the situations that cause `EINTR` for a `sem_t` so I'm not sure what this should be. I was very confused about this as I was leaving today, I'll ask you about it tomorrow.

thehajime commented:
it's a bit off topic though, but helgrind/valgrind-aware custom allocator can be implemented with markers.

http://sourcecodebrowser.com/valgrind/3.6.1/helgrind_8h.html

stfairy commented:
You may read the glibc code to find out implementation details.

pscollins commented:
@thehajime Sounds like just what I was looking for, thanks!

pscollins commented:
I've gone back and cleaned this up a little, and added support for it as a `Kconfig` option. As a very unscientific test, a performance comparison:

```
➜  lkl git:(posix-semaphores) ✗ sudo LKL_HIJACK_NET_TAP=lkl_ptt0 LKL_HIJACK_NET_IP=192.168.13.2  LKL_HIJACK_NET_NETMASK_LEN=24 LKL_LIBRARY_PATH=lib LKL_PRELOAD=liblkl-hijack.so taskset -c 10 bin/lkl-hijack.sh netperf -H 192.168.13.1 -L 192.168.13.2 -t TCP_RR       
MIGRATED TCP REQUEST/RESPONSE TEST from 192.168.13.2 () port 0 AF_INET to 192.168.13.1 () port 0 AF_INET : demo : first burst 0
Local /Remote
Socket Size   Request  Resp.   Elapsed  Trans.
Send   Recv   Size     Size    Time     Rate         
bytes  Bytes  bytes    bytes   secs.    per sec   

16384  87380  1        1       10.00    20624.53   
16384  87380 


➜  lkl git:(posix-semaphores) sudo LKL_HIJACK_NET_TAP=lkl_ptt0 LKL_HIJACK_NET_IP=192.168.13.2  LKL_HIJACK_NET_NETMASK_LEN=24 LKL_LIBRARY_PATH=lib LKL_PRELOAD=liblkl-hijack.so taskset -c 10 bin/lkl-hijack.sh netperf -H 192.168.13.1 -L 192.168.13.2 -t TCP_STREAM
MIGRATED TCP STREAM TEST from 192.168.13.2 () port 0 AF_INET to 192.168.13.1 () port 0 AF_INET : demo
Recv   Send    Send                          
Socket Socket  Message  Elapsed              
Size   Size    Size     Time     Throughput  
bytes  bytes   bytes    secs.    10^6bits/sec  

 87380  16384  16384    10.00    1235.02 (some runs were as low as 1176)

```

vs `pthread_cond_t`:

```
➜  lkl git:(master) sudo LKL_HIJACK_NET_TAP=lkl_ptt0 LKL_HIJACK_NET_IP=192.168.13.2  LKL_HIJACK_NET_NETMASK_LEN=24 LKL_LIBRARY_PATH=lib LKL_PRELOAD=liblkl-hijack.so taskset -c 10 bin/lkl-hijack.sh netperf -H 192.168.13.1 -L 192.168.13.2 -t TCP_STREAM
MIGRATED TCP STREAM TEST from 192.168.13.2 () port 0 AF_INET to 192.168.13.1 () port 0 AF_INET : demo
Recv   Send    Send                          
Socket Socket  Message  Elapsed              
Size   Size    Size     Time     Throughput  
bytes  bytes   bytes    secs.    10^6bits/sec  

 87380  16384  16384    10.00    1185.81  


➜  lkl git:(master) sudo LKL_HIJACK_NET_TAP=lkl_ptt0 LKL_HIJACK_NET_IP=192.168.13.2  LKL_HIJACK_NET_NETMASK_LEN=24 LKL_LIBRARY_PATH=lib LKL_PRELOAD=liblkl-hijack.so taskset -c 10 bin/lkl-hijack.sh netperf -H 192.168.13.1 -L 192.168.13.2 -t TCP_RR
MIGRATED TCP REQUEST/RESPONSE TEST from 192.168.13.2 () port 0 AF_INET to 192.168.13.1 () port 0 AF_INET : demo : first burst 0
Local /Remote
Socket Size   Request  Resp.   Elapsed  Trans.
Send   Recv   Size     Size    Time     Rate         
bytes  Bytes  bytes    bytes   secs.    per sec   

16384  87380  1        1       10.00    15911.79   
16384  87380 
```

So it looks like there is a ~30%  increase on `TCP_RR` and negligible impact on `TCP_STREAM` --- not too bad!

opurdila commented:
This looks really good, performance wise. I this case I think we dont need a LKL option, lets use it by default if we can.

pscollins commented:
@opurdila Done, it tries to use it by default now.

opurdila commented:
Merged with a few changes,  thanks Patrick!
